File: study/2022/22.08.09/HJH/타겟 넘버.java
class Solution {
    static int answer = 0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, 0, target, 0);
        return answer;
    }
    
    public void dfs(int[] numbers, int cnt, int target, int sum){
        if(cnt == numbers.length){
            if(sum == target){
                answer++;
            }
            return;
        }
        dfs(numbers, cnt + 1, target, sum + numbers[cnt]); // 왼쪽
        dfs(numbers, cnt + 1, target, sum - numbers[cnt]); // 오른쪽
    }
}

File: study/2022/22.08.09/EJH/타겟 넘버.py
count = 0


def dfs(n, result, numbers, target):
    global count
    if n == len(numbers):
        if result == target:
            count += 1
        return
    dfs(n + 1, result + numbers[n], numbers, target)
    dfs(n + 1, result - numbers[n], numbers, target)


def solution(numbers, target):
    answer = 0
    global count
    count = 0
    dfs(0, 0, numbers, target)
    answer = count
    return answer



print(solution([1, 1, 1, 1, 1], 3))
print(solution([4, 1, 2, 1], 4))